test/partition.test.ts

```diff
@@ -0,0 +1,14 @@
+import { expectType } from 'tsd';
+import { partition } from '../types/partition';
+
+// partition(fn, list)
+// fn: (v) => v is T
+expectType<[true[], false[]]>(partition((v): v is true => v, [true, false]));
+// fn: (v) => bool
+expectType<[boolean[], boolean[]]>(partition((v) => v, [true, false]));
+
+// partition(fn)(list)
+// fn: (v) => v is T
+expectType<[true[], false[]]>(partition((v): v is true => !!v)([true, false]));
+// fn: (v) => bool
+expectType<[boolean[], boolean[]]>(partition((v) => !!v)([true, false]));
```

Comment:
In my other comment, the bits about `L extends T = T` and `T & U`. If those are necessary for reasons I'm just not aware of, the best way to show that would be through tests. If you can create tests for the specific situation that would require those and would fail otherwise, that would help greatly